### DISCUSSION TITLE
Embedding a mpv/vlc video player inside a window

### DISCUSSION THREAD

**Discussion by prayag17** [tauri-apps/tauri#6343] [Q&A]
Hello, can you embed mpv video player inside a window?
I want to have a webview with a transparent background and a canvas underneath this webview which displays videos with the help of mpv. 
I looked at some examples like https://github.com/vgarleanu/lumiere which don't use tauri but might be able to implement it for tauri but I cannot find a way to achieve this.
I know that there is a working mpv binding for rust but its documentation is vague.

There is also vlc-rc but that to doesn't have a good documentation

**Comment by vgarleanu** [comment]
I think a possible avenue is embedding a video player's window in a tauri window.

**Reply by CaseSens** [reply]
This would be the simplest, if all you're looking to do is play videos. As for the transparent background, Tauri offers transparent backgrounds using [this solution](https://docs.rs/tauri/latest/tauri/window/struct.WindowBuilder.html#method.transparent). Keep in mind this will only work on non-MacOs environments. For MacOs, you can see the official tauri documentation [here](https://v2.tauri.app/learn/window-customization/#macos-transparent-titlebar-with-custom-window-background-color)

**Comment by Dr-42** [comment]
Has there been a solution to doing this? I too want to mpv into my tauri window for video-playback

**Reply by prayag17** [reply]
I was able to embed the CLI version of mpv in the Tauri application using RawWindowHandle, tho this method is pretty useless since you don't get control over the running mpv instance. 

You can refer to my discord thread:
https://discord.com/channels/616186924390023171/1215223447693692958

**Reply by s074** [reply]
You can use libmpv to do the same, and you will have access to pass commands to the mpv player. Only challenge with that route is that the rust bindings for the library are lacking. There's this one: https://github.com/kohsine/libmpv2-rs which supports libmpv v2, but hasn't been updated in months. You can probably do something like:
```
let mpv = Mpv::with_initializer(|init| {
    init.set_property("vo", "gpu")?;
    init.set_option("wid", window_id as i64)?;
    Ok(())
})
.unwrap();

mpv.command("loadfile", &["http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4", "replace"]).unwrap(); 
```

I was able to get vlc-rs working with it, which might be the way to go because the bindings are maintained by vlc people and you can get support from their Discord (they were very helpful to me). I had to render the webview in a child window though because vlc kept rendering on top of it if done in the same window.

**Reply by nini22P** [reply]
> I was able to embed the CLI version of mpv in the Tauri application using RawWindowHandle, tho this method is pretty useless since you don't get control over the running mpv instance.
> 
> You can refer to my discord thread: https://discord.com/channels/616186924390023171/1215223447693692958

You can control mpv playback via JSON IPC
https://github.com/nini22P/mpv-tauri

**Comment by Dr-42** [comment]
Anyone looking for a solution, No I didn't embed mpv.

I created a [server](https://github.com/Dr-42/vserve) with backend and video frontend that live transcodes and streams it to the browser video player

**Comment by soleuniverse101** [comment]
If someone is looking to do this, I was able to create a child WebviewWindow and run libmpv in it using [tauri-plugin-libmpv](https://github.com/nini22P/tauri-plugin-libmpv).

Downside is that I'm faking it and the window is not truly embedded, basically when the parent moves or resizes I reset the position and size of the libmpv window to align it with an HTML div (whose coordinates I get through `getBoundingClientRect()`). Also all commands have to be set manually in the main WebView to then forward them to libmpv through commands and property changes (I'm using [hotkeys-js](https://github.com/jaywcjlove/hotkeys-js) for that).

One way that could work (on the word of ChatGPT...) would be to configure libmpv to render in a GPU texture that would then be displayed in the Tauri WebView through wgpu and/or a canvas. I won't go that route for now since it seems difficult but if someone wants to enable true embedding, as an HTML component, they can explore that way. If someone accomplishes this, I would very much appreciate to be notified!

**Comment by blurymind** [comment]
webkit gtk is kind of broken with mp4 streaming playback at the moment. I doesnt play the video to the end, even when you do give it buffer headers
